src/core/metadata/qgslayermetadata.cpp

```diff
@@ -273,6 +278,312 @@ void QgsLayerMetadata::readFromLayer( const QgsMapLayer *layer )
+      QDomText keywordText = document.createTextNode( kw );
+      keyword.appendChild( keywordText );
+      keywordsElement.appendChild( keyword );
+    }
+    metadataElement.appendChild( keywordsElement );
+  }
+
+  // fees
+  QDomElement fees = document.createElement( QStringLiteral( "fees" ) );
+  QDomText feesText = document.createTextNode( mFees );
+  fees.appendChild( feesText );
+  metadataElement.appendChild( fees );
+
+  // constraints
+  for ( const QgsLayerMetadata::Constraint constraint : constraints() )
+  {
+    QDomElement constraintElement = document.createElement( QStringLiteral( "constraints" ) );
+    constraintElement.setAttribute( QStringLiteral( "type" ), constraint.type );
+    QDomText constraintText = document.createTextNode( constraint.constraint );
+    constraintElement.appendChild( constraintText );
+    metadataElement.appendChild( constraintElement );
+  }
+
+  // rights
+  for ( const QString right : rights() )
+  {
+    QDomElement rightElement = document.createElement( QStringLiteral( "rights" ) );
+    QDomText rightText = document.createTextNode( right );
+    rightElement.appendChild( rightText );
+    metadataElement.appendChild( rightElement );
+  }
+
+  // encoding
+  QDomElement encoding = document.createElement( QStringLiteral( "encoding" ) );
+  QDomText encodingText = document.createTextNode( mEncoding );
+  encoding.appendChild( encodingText );
+  metadataElement.appendChild( encoding );
+
+  // crs
+  QDomElement crsElement = document.createElement( QStringLiteral( "crs" ) );
+  mCrs.writeXml( crsElement, document );
+  metadataElement.appendChild( crsElement );
+
+  // contact
+  for ( const QgsLayerMetadata::Contact contact : mContacts )
+  {
+    QDomElement contactElement = document.createElement( QStringLiteral( "contact" ) );
+    QDomElement nameElement = document.createElement( QStringLiteral( "name" ) );
+    QDomElement organizationElement = document.createElement( QStringLiteral( "organization" ) );
+    QDomElement positionElement = document.createElement( QStringLiteral( "position" ) );
+    QDomElement voiceElement = document.createElement( QStringLiteral( "voice" ) );
+    QDomElement faxElement = document.createElement( QStringLiteral( "fax" ) );
+    QDomElement emailElement = document.createElement( QStringLiteral( "email" ) );
+    QDomElement roleElement = document.createElement( QStringLiteral( "role" ) );
+
+    QDomText nameText = document.createTextNode( contact.name );
+    QDomText orgaText = document.createTextNode( contact.organization );
+    QDomText positionText = document.createTextNode( contact.position );
+    QDomText voiceText = document.createTextNode( contact.voice );
+    QDomText faxText = document.createTextNode( contact.fax );
+    QDomText emailText = document.createTextNode( contact.email );
+    QDomText roleText = document.createTextNode( contact.role );
+
+    for ( QgsLayerMetadata::Address oneAddress : contact.addresses )
```

Comment:
missing reference

Reply:
@nyalldawson I tried, but it gives me:
```
/Users/etienne/dev/cpp/QGIS/src/core/metadata/qgslayermetadata.cpp:560: error: binding value of type 'const QgsLayerMetadata::Address' to reference to type 'QgsLayerMetadata::Address' drops 'const' qualifier
    for ( QgsLayerMetadata::Address& oneAddress : contact.addresses )
                                     ^          ~
```
 Any hint?